src/wp-includes/blocks.php

```diff
@@ -1596,10 +1596,26 @@ function make_after_block_visitor( $hooked_blocks, $context, $callback = 'insert
  *
  * @since 5.3.1
  *
- * @param array $block_attributes Attributes object.
+ * @param array       $block_attributes Attributes object.
+ * @param string|null $block_name       Block name. Null if the block name is unknown. Optional.
  * @return string Serialized attributes.
  */
-function serialize_block_attributes( $block_attributes ) {
+function serialize_block_attributes( $block_attributes, $block_name = null ) {
+	foreach ( $block_attributes as $attribute => $value ) {
+		if ( is_array( $value ) && empty( $value ) ) {
+			// An empty `array()` is encoded as `[]` in JSON. However, it's possible
+			// that the attribute type is really an object (associative array in PHP),
+			// so we need to check for that.
+			$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block_name );
+			if ( $block_type && isset( $block_type->attributes[ $attribute ] ) ) {
+				$attribute_type = $block_type->attributes[ $attribute ]['type'];
```

Comment:
Another nit: `type` isn't require, see JSON schema:

https://github.com/WordPress/gutenberg/blob/3eea2dfb6a01e1e8282d9c4dfc2e84ba81429e0c/schemas/json/block.json#L192-L195

Devs can use `enum` or `type`.

It would be nice to include check for `type` in the existing `isset` check.

Reply:
27512a1145dce46e25e3575ebab4f9ccf1b52fb1